CorsixTH/Src/iso_fs.cpp

```diff
@@ -70,33 +75,187 @@ constexpr int max_directory_depth = 16;
 /// size is 2048.
 constexpr size_t min_sector_size = 2048;
 
-template <class T> inline T read_native_int(const uint8_t *p)
+/// Offset of the sector size from the primary volume descriptor
+constexpr size_t sector_size_offset = 128;
+
+/// Offset of the root directory entry from the primary volume descriptor
+constexpr ptrdiff_t root_directory_offset = 156;
+
+/// The root directory entry is a fixed size.
+constexpr size_t root_directory_entry_size = 34;
+
+/// ISO 9660 has a 32kb reserve area at the beginning of the file formal
+/// e.g. boot information.
+constexpr uint32_t first_filesystem_sector = 16;
+
+void trim_identifier_version(const uint8_t* sIdent, uint8_t& iLength)
 {
-    // ISO 9660 commonly encodes multi-byte integers as little endian followed
-    // by big endian. Note that the first byte of iEndianness will be a zero on
-    // little endian systems, and a one on big endian.
-    static const uint16_t iEndianness = 0x0100;
-    return reinterpret_cast<const T*>(p)[*reinterpret_cast<const uint8_t*>(&iEndianness)];
+    for(uint8_t i = 0; i < iLength; ++i)
+    {
+        if(sIdent[i] == ';')
+        {
+            iLength = i;
+            return;
+        }
+    }
 }
 
-} // namespace
+char normalise(char c) {
+    if (c == '_') {
+        return '-';
+    } else if ('a' <= c && c <= 'z') {
+        return static_cast<char>(c - 'a' + 'A');
+    } else {
+        return c;
+    }
+}
 
-iso_filesystem::iso_filesystem()
-{
-    raw_file = nullptr;
-    error = nullptr;
-    files = nullptr;
-    file_count = 0;
-    file_table_size = 0;
-    path_seperator = '\\';
+std::string normalise(const uint8_t* start, size_t length) {
+    std::string ret;
+    const uint8_t* p = start;
+    for (size_t i = 0; i < length; i++) {
+        ret.push_back(normalise(static_cast<char>(*p)));
+        ++p;
+    }
+    return ret;
 }
 
+std::string normalise(const char* str) {
+    std::string ret;
+    const char* p = str;
+    while (*p != '\0') {
+        ret.push_back(normalise(*p));
+        ++p;
+    }
+    return ret;
+}
+
+class iso_file_entry {
+public:
+    iso_file_entry()=default;
+    iso_file_entry(const uint8_t* b) {
+        data_sector = bytes_to_uint32_le(b + file_sector_offset);
+        data_length = bytes_to_uint32_le(b + file_data_length_offset);
+        flags = b[file_flags_offset];
+        uint8_t filename_length = b[filename_length_offset];
+        trim_identifier_version(b + filename_offset, filename_length);
+        filename = normalise(b + filename_offset, filename_length);
+    }
+
+    uint32_t data_sector;
+    uint32_t data_length;
+    uint8_t flags;
+    std::string filename;
+};
+
+class iso_directory_iterator final {
+    using iterator_category = std::input_iterator_tag;
+    using value_type = const iso_file_entry;
+    using difference_type = ptrdiff_t;
+    using pointer = const iso_file_entry*;
+    using reference = const iso_file_entry&;
```

Comment:
new spiffy type syntax, apparently :)

Reply:
C++ 11 version of a typedef yes. This is the standard C++ 17 (and higher) way of declaring an iterator, since std::iterator is deprecated. Works in C++14 so why not.